### PR TITLE
feat: address page support `node-connect` mode

### PR DESCRIPTION
**Pull request by PainterPuppets**
resolved: https://github.com/Magickbase/ckb-explorer-public-issues/issues/526

### CODE REVIEW

File: src/pages/Address/styled.tsx [hunk 1]
```diff
@@ -151,3 +151,29 @@ export const AddressUDTAssetsPanel = styled.div`
     border-top: 1px solid #f5f5f5;
   }
 `
+
+export const CardFooterPanel = styled.div`
```
Comment (Keith-CY) [inline src/pages/Address/styled.tsx:155]: We can use less styled component

**Review by Keith-CY — APPROVED** [review]

**Comment by FrederLu** [comment]
https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy
image
image


https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl

If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.

**Comment by PainterPuppets** [comment]
> https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy image image
> 
> https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl
> 
> 1、If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.
> 
> image image
> 2、Missing transaction statistics and `Pending Transactions (0)` information.
> 
> image image
> 3、There is a problem with the data output[3] in the `Node` mode.
> 
> https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9 image
> 
> image
> 4、The transaction address is missing icons (->/O) on the left and right.

2. The `fetch tx` by `node-rpc` are waterfall streaming, there's no way to know how many tx there are.

About peding tx: the `node-rpc` can only fetch all the tx in the tx pool, there's no way to filter and aggregate by address, and if we filter and aggregate on the front-end, there's a performance issue if there's more tx (for example, when build inscriptions before)

4. These two icons show where the cell was generated from and the state of consumption, there is no way for the `node-rpc` to provide this feature, so it is removed

**Comment by PainterPuppets** [comment]
> > https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy image image
> > https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl
> > 1、If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.
> > image image
> > 2、Missing transaction statistics and `Pending Transactions (0)` information.
> > image image
> > 3、There is a problem with the data output[3] in the `Node` mode.
> > https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9 image
> > image
> > 4、The transaction address is missing icons (->/O) on the left and right.
> 
> 2. The `fetch tx` by `node-rpc` are waterfall streaming, there's no way to know how many tx there are.
> 
> About peding tx: the `node-rpc` can only fetch all the tx in the tx pool, there's no way to filter and aggregate by address, and if we filter and aggregate on the front-end, there's a performance issue if there's more tx (for example, when build inscriptions before)
> 
> 4. These two icons show where the cell was generated from and the state of consumption, there is no way for the `node-rpc` to provide this feature, so it is removed

**Comment by Keith-CY** [comment]
> > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy) image image
> > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl)
> > 1、If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.
> > image image
> > 2、Missing transaction statistics and `Pending Transactions (0)` information.
> > image image
> > 3、There is a problem with the data output[3] in the `Node` mode.
> > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9) image
> > image
> > 4、The transaction address is missing icons (->/O) on the left and right.
> 
> 2. The `fetch tx` by `node-rpc` are waterfall streaming, there's no way to know how many tx there are.
> 
> About peding tx: the `node-rpc` can only fetch all the tx in the tx pool, there's no way to filter and aggregate by address, and if we filter and aggregate on the front-end, there's a performance issue if there's more tx (for example, when build inscriptions before)
> 
> 3. This cell is an xudt cell, and according to normal logic it should show the udt amount, so it feels logically fine, maybe we should show it as `XUDT amount (ckb capacity)`
> 4. These two icons show where the cell was generated from and the state of consumption, there is no way for the `node-rpc` to provide this feature, so it is removed

For point 4. showing the `from` of `inputs` and `to` of `outputs`
It's partially feasible with RPC https://github.com/nervosnetwork/ckb/tree/develop/rpc#method-get_live_cell
If the RPC returns `null`, the `output` has been used by some transaction. The live/dead status can be determined.

**Comment by FrederLu** [comment]
> > 1、If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.

Now the paging logic is OK, but should we also display 25 data items on a single page to keep it consistent with the non-`Node` mode? Now there are 10 data items on a single page.


> 2. The `fetch tx` by `node-rpc` are waterfall streaming, there's no way to know how many tx there are.
> 
> About peding tx: the `node-rpc` can only fetch all the tx in the tx pool, there's no way to filter and aggregate by address, and if we filter and aggregate on the front-end, there's a performance issue if there's more tx (for example, when build inscriptions before)
> 
> 3. This cell is an xudt cell, and according to normal logic it should show the udt amount, so it feels logically fine, maybe we should show it as `XUDT amount (ckb capacity)`
> 4. These two icons show where the cell was generated from and the state of consumption, there is no way for the `node-rpc` to provide this feature, so it is removed

The third question is how do we display the data?  @Keith-CY

**Comment by Keith-CY** [comment]
> > > 1、If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.
> 
> Now the paging logic is OK, but should we also display 25 data items on a single page to keep it consistent with the non-`Node` mode? Now there are 10 data items on a single page. @PainterPuppets cc @Keith-CY
> 
> > 2. The `fetch tx` by `node-rpc` are waterfall streaming, there's no way to know how many tx there are.
> > 
> > About peding tx: the `node-rpc` can only fetch all the tx in the tx pool, there's no way to filter and aggregate by address, and if we filter and aggregate on the front-end, there's a performance issue if there's more tx (for example, when build inscriptions before)
> > 
> > 3. This cell is an xudt cell, and according to normal logic it should show the udt amount, so it feels logically fine, maybe we should show it as `XUDT amount (ckb capacity)`
> > 4. These two icons show where the cell was generated from and the state of consumption, there is no way for the `node-rpc` to provide this feature, so it is removed
> 
> The third question is how do we display the data? @Keith-CY

The `outputs[3]` cell has `xudt type script` so it should be displayed with `xudt amount`. The preview is correct while the display of testnet is incorrect

**Comment by PainterPuppets** [comment]
> > > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy) image image
> > > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl)
> > > 1、If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.
> > > image image
> > > 2、Missing transaction statistics and `Pending Transactions (0)` information.
> > > image image
> > > 3、There is a problem with the data output[3] in the `Node` mode.
> > > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9) image
> > > image
> > > 4、The transaction address is missing icons (->/O) on the left and right.
> > 
> > 
> > 
> > 2. The `fetch tx` by `node-rpc` are waterfall streaming, there's no way to know how many tx there are.
> > 
> > About peding tx: the `node-rpc` can only fetch all the tx in the tx pool, there's no way to filter and aggregate by address, and if we filter and aggregate on the front-end, there's a performance issue if there's more tx (for example, when build inscriptions before)
> > 
> > 3. This cell is an xudt cell, and according to normal logic it should show the udt amount, so it feels logically fine, maybe we should show it as `XUDT amount (ckb capacity)`
> > 4. These two icons show where the cell was generated from and the state of consumption, there is no way for the `node-rpc` to provide this feature, so it is removed
> 
> For point 4. showing the `from` of `inputs` and `to` of `outputs` It's partially feasible with RPC https://github.com/nervosnetwork/ckb/tree/develop/rpc#method-get_live_cell If the RPC returns `null`, the `output` has been used by some transaction. The live/dead status can be determined.

In this case the list will send more requests, for example, a list with 10 tx might have 30 cells and require 30 requests to get the live status. I'm not so sure about this

**Comment by Keith-CY** [comment]
> > > > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzq2p8k2ymthe6sltf55w8zefqd7wsz0a06qa6g03pkrd22g8pd42qtaqqqqqyqqqqq9jqqqqpwsqqqqfyqqqqqsqqqqqvqqqqqrzqqqqzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqg5qqqqptz8d3wnv5vqyz53mgjs4m6afapasylrkpsqqqqqmcggt45w3qxrfkxhl3wykrw75tfchkt5f9vfzl4d7355yy8r5xql6ugfy) image image
> > > > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qyqrpp98atnkx060dh3mhj7txvxgvj3rszeqd4lwxl)
> > > > 1、If paging is not required, the `Load more` button at the bottom should not be displayed, and `No more data` should be displayed directly. Now the current transaction data is 5, and the `Load more` button is displayed, but the data does not increase after clicking it.
> > > > image image
> > > > 2、Missing transaction statistics and `Pending Transactions (0)` information.
> > > > image image
> > > > 3、There is a problem with the data output[3] in the `Node` mode.
> > > > [ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9](https://ckb-explorer-frontend-in-magickbase-repo-epkr0oa3n-magickbase.vercel.app/address/ckt1qzda0cr08m85hc8jlnfp3zer7xulejywt49kt2rr0vthywaa50xwsqfssjn74emr8a8kmcame09nxryxfg3cpvsz27yt9) image
> > > > image
> > > > 4、The transaction address is missing icons (->/O) on the left and right.
> > > 
> > > 
> > > 
> > > 2. The `fetch tx` by `node-rpc` are waterfall streaming, there's no way to know how many tx there are.
> > > 
> > > About peding tx: the `node-rpc` can only fetch all the tx in the tx pool, there's no way to filter and aggregate by address, and if we filter and aggregate on the front-end, there's a performance issue if there's more tx (for example, when build inscriptions before)
> > > 
> > > 3. This cell is an xudt cell, and according to normal logic it should show the udt amount, so it feels logically fine, maybe we should show it as `XUDT amount (ckb capacity)`
> > > 4. These two icons show where the cell was generated from and the state of consumption, there is no way for the `node-rpc` to provide this feature, so it is removed
> > 
> > 
> > For point 4. showing the `from` of `inputs` and `to` of `outputs` It's partially feasible with RPC [nervosnetwork/ckb@`develop`/rpc#method-get_live_cell](https://github.com/nervosnetwork/ckb/tree/develop/rpc?rgh-link-date=2024-07-05T09%3A18%3A47Z#method-get_live_cell) If the RPC returns `null`, the `output` has been used by some transaction. The live/dead status can be determined.
> 
> In this case the list will send more requests, for example, a list with 10 tx might have 30 cells and require 30 requests to get the live status. should we do this?

The node-connection mode works as a fallback in case the server is down, or a solution to connect to the local node, I think the traffic is tolerant, especially when it's connected to a local node.

Besides, the performance can be optimized by batching the RPC calls.

**Review by FrederLu — APPROVED** [review]

**Pull request merged**